solvers/geomloss.py

```diff
@@ -0,0 +1,112 @@
+from benchopt import BaseSolver, safe_import_context
+from benchopt.stopping_criterion import SufficientProgressCriterion
+
+# Protect the import with `safe_import_context()`. This allows:
+# - skipping import to speed up autocompletion in CLI.
+# - getting requirements info when all dependencies are not installed.
+with safe_import_context() as import_ctx:
+    from geomloss.sinkhorn_divergence import log_weights, sinkhorn_loop
+    from geomloss.sinkhorn_samples import cost_routines, softmin_tensorized
+    import jax.numpy as jnp
+    import numpy as np
+    import ott
+    from ott.geometry import pointcloud
+    from ott.problems.linear import linear_problem
+    from ott.solvers.linear.sinkhorn import SinkhornOutput
+    import torch
+
+
+# The benchmark solvers must be named `Solver` and
+# inherit from `BaseSolver` for `benchopt` to work properly.
+class Solver(BaseSolver):
+
+    # Name to select the solver in the CLI and to display the results.
+    name = 'GeomLoss'
+
+    install_cmd = 'conda'
+    requirements = ['torch', 'pykeops', 'pip:geomloss', 'ott-jax']
+
+    # List of parameters for the solver. The benchmark will consider
+    # the cross product for each key in the dictionary.
+    parameters = {
+        'reg': [1e-2, 1e-1],
+    }
+
+    stopping_criterion = SufficientProgressCriterion(patience=50)
+
+    def set_objective(self, x, a, y, b):
+        # Convert problem into jax array with int32 for jitted computations.
+        x_jax, y_jax, a_jax, b_jax = map(
+            lambda x: jnp.array(x), (x, y, a, b)
+        )
+        self.ot_prob = linear_problem.LinearProblem(
+            pointcloud.PointCloud(
+                x_jax, y_jax, epsilon=self.reg,
+                cost_fn=ott.geometry.costs.SqPNorm(p=2)
+            ), a_jax, b_jax,
+        )
+        self.x, self.a, self.y, self.b = [
+            torch.from_numpy(t).float()[None] for t in (x, a, y, b)
+        ]
+        # put all tensors on GPU if available
+        if torch.cuda.is_available():
+            self.x, self.a, self.y, self.b = [
+                t.cuda() for t in (self.x, self.a, self.y, self.b)
+            ]
```

Comment:
```suggestion

        # Store the problem in torch to use GeomLoss.
        # Use the GPU when it is available.
        device = 'cuda' if torch.cuda.is_available() else 'cpu'
        self.x, self.a, self.y, self.b = [
            torch.from_numpy(t).float().to(device=device)[None]
            for t in (x, a, y, b)
        ]
```

Reply:
For the GPU, maybe a good idea would be to set a parameter for this? So we can compare both implementations?

Reply:
Why not, will implement